index.js

```diff
@@ -835,6 +871,68 @@ class Batch {
   }
 }
 
+class Watcher extends EventEmitter {
+  constructor (bee, range) {
+    super()
+
+    this.bee = bee
+    this.core = bee.core
+
+    this.running = false
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
+    this.stream = null
+
+    this._onappend = this._onappend.bind(this)
+    this._onappend = debounceify(this._onappend)
+
+    // this.on('error', noop)
+  }
+
+  async _onappend () {
+    this.running = true
+
+    try {
+      await this._run()
+    } catch (err) {
+      this.emit('error', err)
+      this.destroy()
+    } finally {
+      this.running = false
+    }
+  }
+
+  async _run () {
+    if (this.destroyed) return
+
+    const snapshot = this.bee.snapshot()
+    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+
+    try {
+      for await (const data of this.stream) { // eslint-disable-line
+        this.emit('change', snapshot.version, this.latestDiff)
+        break
+      }
+    } catch (err) {
+      if (this.destroyed && err.message === 'Stream was destroyed') return
+      throw err
+    }
+
+    this.latestDiff = snapshot.version
+  }
+
+  destroy () {
+    if (this.destroyed) return
+    this.destroyed = true
+
+    if (this.stream) this.stream.destroy()
```

Comment:
wait for the stream to emit close here

Reply:
Should I use `destroying`?
https://github.com/holepunchto/hyperbee/blob/bb99190279b2263bef26848a3b318f074b543f8a/index.js#L933
With `destroyed` also works, but sounds like if `destroying` is `true` then no need to `destroy()` the stream

Another version, this sounds even more precise but it could be unnecessary:
```js
if (this.stream && !this.stream.destroyed) {
    if (!this.stream.destroying) this.stream.destroy()
    this.stream.once('close', emitClose)
  } else {
  setImmediate(emitClose)
}
```